### PR TITLE
fix(inductor): harden copy_from_d2d offset guard (symbolic + stick-dim)

### PR DESCRIPTION
**Pull request by thoangtrvn**
#### What type of PR is this?

- [x] bug
- [ ] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:

Follow-up to #3237. That PR added _validate_reoffset_supported to reject a d2d-copy storage_offset that is not a whole number of sticks. Two gaps in that guard could still lead to silent wrong data; close both, keeping the same lowering-time fail-fast behavior.

1. Symbolic offset. A non-int (symbolic) offset was let through with a bare `return`. It would then reach superdsc, whose as_coeff_Add()[0] constant term is 0 for a bare symbol, baking a zero byte-offset and silently reading from the storage base — the exact #3236 bug. specialize_int is meant to bake every offset as a concrete int, so this branch should be unreachable; make it raise Unsupported loudly rather than risk corruption if that invariant breaks.

2. Offset inside the stick dimension. offset % elems_per_stick == 0 accepts a column narrow (e.g. x.narrow(1, 64, 64) -> stride (128, 1), offset 64) whose offset starts partway into a row; superdsc does not split a stick-dim offset, so the copy silently reads wrong data (measured: got 0.0, expected 64.0). This is detectable from the view's OWN stride: offset % stride[-2] != 0 means the offset is not on an outer-dimension boundary. Reject it here instead of miscompiling. (The earlier docstring's claim that detecting this needs the base-storage layout was wrong — the view stride reveals it.)

The underlying stick-dim copy is still unsupported (fails cleanly), tracked for the follow-up work.

Tests: flip test_column_slice_inner_offset to _raises (was expectedFailure / silent wrong, now a clean raise), and add TestValidateReoffsetUnit for the symbolic branch (the one guard path with no e2e equivalent).

Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).

#### Which issue(s) this PR is related to:

Issue #3236

### CODE REVIEW

File: tests/inductor/test_copy_from_d2d_offsets.py [hunk 1]
```diff
@@ -99,24 +102,25 @@ def test_copy_into_sliced_dst(self):
...
+        underlying stick-dim copy is still unsupported (tracked for the follow-up
+        PR); this test asserts the failure is clean, not silent."""
         x = torch.arange(2 * 128, dtype=DTYPE, device=DEVICE).reshape(2, 128)
-        # columns [64:128) -> nonzero offset within a row
-        out = x.narrow(1, 64, 64).clone()
-        torch.testing.assert_close(out.cpu(), x.cpu()[:, 64:128])
+        # columns [64:128) -> offset 64 lands inside the row (stride 128)
+        with self.assertRaises(Exception) as cm:
+            x.narrow(1, 64, 64).clone()
+        self.assertIn("stick", str(cm.exception).lower())
```
Comment (Vikram123-byte) [inline tests/inductor/test_copy_from_d2d_offsets.py]: Assert the specific validation failures.

These broad Exception checks can pass on unrelated compiler errors. Can we match the distinctive guard messages so the regressions prove the intended branches executed.

File: tests/inductor/test_copy_from_d2d_offsets.py [hunk 2]
```diff
@@ -238,5 +242,35 @@ def test_transpose_varying_offsets_loop(self):
...
+        """A symbolic (non-int) offset must raise, not silently bake 0."""
+        import sympy
+
+        from torch_spyre._inductor.lowering import _validate_reoffset_supported
+
+        with self.assertRaises(Exception) as cm:
+            _validate_reoffset_supported(
+                self._FakeLayout(DTYPE, [64, 1]), sympy.Symbol("s0")
+            )
+        self.assertIn("symbolic", str(cm.exception).lower())
```
Comment (Vikram123-byte) [inline tests/inductor/test_copy_from_d2d_offsets.py]: https://github.com/torch-spyre/torch-spyre/pull/3285/changes#r3630720976

File: torch_spyre/_inductor/lowering.py [hunk 3]
```diff
@@ -913,6 +938,23 @@ def _validate_reoffset_supported(layout, offset) -> None:
             "coordinate. Re-slice on a stick-aligned boundary (a multiple of "
             f"{eps} elements), or copy the full (unsliced) tensor."
         )
+    # Offset must land on an outer-dimension boundary; a remainder smaller than
+    # the innermost dim's stride means it falls inside the stick dim (a column
+    # narrow), which superdsc bakes incorrectly. Detectable from the view stride.
+    try:
+        stride = [int(s) for s in layout.stride]
+    except (TypeError, ValueError):
+        return  # symbolic strides: the stick-aligned check above still applied
```
Comment (Vikram123-byte) [inline torch_spyre/_inductor/lowering.py]: Reject symbolic strides rather than bypassing the boundary check.

A concrete, stick-aligned offset still cannot be proven safe against a symbolic stride[-2]; returning here can re-enable the silent column-offset miscompile this guard is meant to prevent. Fail loudly, as the symbolic-offset branch does.

Comment (thoangtrvn) [reply torch_spyre/_inductor/lowering.py]: Agreed. Thanks @Vikram123-byte .

File: tests/inductor/test_copy_from_d2d_offsets.py (hunk 2 above)
Comment (thoangtrvn) [reply tests/inductor/test_copy_from_d2d_offsets.py]: Updated.

File: tests/inductor/test_copy_from_d2d_offsets.py (hunk 1 above)
Comment (thoangtrvn) [reply tests/inductor/test_copy_from_d2d_offsets.py]: Updated.

File: torch_spyre/_inductor/lowering.py [hunk 4]
```diff
@@ -913,6 +938,33 @@ def _validate_reoffset_supported(layout, offset) -> None:
...
+        # stick-aligned offset could still fall inside the stick dim and silently
+        # miscompile (the column-offset case). Fail loudly rather than bypass the
+        # boundary check, mirroring the symbolic-offset branch above.
+        raise Unsupported(
+            "spyre::copy_from_d2d of a sliced view received a symbolic stride "
+            f"({list(layout.stride)!r}); a concrete stride is required to prove "
+            f"the storage_offset {off} lands on an outer-dimension boundary and "
+            "not inside the stick dimension. A symbolic stride cannot rule out "
+            "the silent column-offset miscompile, so it is rejected."
+        )
```
Comment (Vikram123-byte) [inline torch_spyre/_inductor/lowering.py]: Stride conversion is scoped too broadly — can spuriously reject valid offsets.

[int(s) for s in layout.stride] converts every stride entry, but only stride[-2] is used by the boundary check. Two consequences:

1. For rank < 2 tensors, a symbolic single stride entry now raises Unsupported even though the boundary check (len(stride) >= 2) never applies to them.
2. For rank ≥ 3 tensors, a symbolic stride on an unrelated outer/batch dim (common with dynamic shapes) will raise Unsupported even when stride[-2] itself is concrete and the actual invariant being protected holds fine.

This goes beyond the documented "safe failure direction" (which targets the stick-dim boundary specifically) and can needlessly break compiles unrelated to the column-offset miscompile risk this guard is meant to catch.



```suggestion
    # Offset must land on an outer-dimension boundary; a remainder smaller than
    # the innermost dim's stride means it falls inside the stick dim (a column
    # narrow), which superdsc bakes incorrectly. Detectable from the view stride.
    if len(layout.stride) >= 2:
        try:
            stride_neg2 = int(layout.stride[-2])
        except (TypeError, ValueError):
            # A symbolic stride[-2] cannot prove off % stride[-2] == 0, so a concrete
            # stick-aligned offset could still fall inside the stick dim and silently
            # miscompile (the column-offset case). Fail loudly rather than bypass the
            # boundary check, mirroring the symbolic-offset branch above.
            raise Unsupported(
                "spyre::copy_from_d2d of a sliced view received a symbolic stride "
                f"({layout.stride[-2]!r}); a concrete stride is required to prove "
                f"the storage_offset {off} lands on an outer-dimension boundary and "
                "not inside the stick dimension. A symbolic stride cannot rule out "
                "the silent column-offset miscompile, so it is rejected."
            )
        if stride_neg2 and off % stride_neg2 != 0:
            raise Unsupported(
                "spyre::copy_from_d2d of a sliced view requires an offset on an "
                f"outer-dimension boundary: storage_offset {off} leaves a remainder "
                f"of {off % stride_neg2} within the innermost dimension "
                f"(stride {stride_neg2}), i.e. it falls inside the stick dimension "
                "and cannot be baked into the kernel coordinate. This happens for a "
                "column slice (narrow on the last dim); copy the full rows instead, "
                "or make the sliced region contiguous first."
            )
```

File: torch_spyre/_inductor/lowering.py [hunk 5]
```diff
@@ -913,6 +938,33 @@ def _validate_reoffset_supported(layout, offset) -> None:
...
+    if len(stride) >= 2 and stride[-2] and off % stride[-2] != 0:
```
Comment (Vikram123-byte) [inline torch_spyre/_inductor/lowering.py:962]: nitpick: stride[-2] falsy bypass — worth a comment or explicit note.

When stride[-2] is 0 (e.g. a broadcast dim), the boundary check is silently skipped, so an offset that would otherwise land inside a stick could pass unnoticed. Likely a rare edge case for D2D copy views here, but the intent (guard against silent miscompile) would benefit from an explicit comment stating why zero stride is safe to skip, or from tightening the condition.

File: tests/inductor/test_copy_from_d2d_offsets.py [hunk 6]
```diff
@@ -238,5 +243,56 @@ def test_transpose_varying_offsets_loop(self):
...
+
+        # offset 128 is stick-aligned (128 % 64 == 0) so it clears branch (1);
+        # the symbolic stride[-2] then cannot prove the boundary condition.
+        with self.assertRaises(Unsupported) as cm:
+            _validate_reoffset_supported(
+                self._FakeLayout(DTYPE, [sympy.Symbol("s0"), 1]), 128
+            )
+        self.assertIn("symbolic", str(cm.exception).lower())
+
+
```
Comment (Vikram123-byte) [inline tests/inductor/test_copy_from_d2d_offsets.py:319]: Solid direct coverage, but doesn't catch the over-broad stride-conversion scope.

test_symbolic_stride_raises uses a 2-element stride [sympy.Symbol("s0"), 1], where the symbolic entry happens to be stride[-2] itself. It doesn't exercise the case where an unrelated stride entry (e.g. rank ≥ 3 with symbolic outer stride, concrete stride[-2]) would currently also spuriously raise given the current implementation converts the entire stride list. Consider adding a case with stride=[sympy.Symbol("s0"), 64, 1] (offset stick-aligned, stride[-2] concrete) to lock in the fix suggested for lowering.py lines 941-957.

File: torch_spyre/_inductor/lowering.py (hunk 3 above)
Comment (thoangtrvn) [reply torch_spyre/_inductor/lowering.py]: Actually, after the rebase, I realized that a main's recent commits added automatic-dynamic shape compilation to the D2D-copy path. Under dynamic shapes stride[-2] becomes a symbol, which made the column-narrow boundary check (offset % stride[-2]) undecidable.

File: torch_spyre/_inductor/lowering.py (hunk 4 above)
Comment (thoangtrvn) [reply torch_spyre/_inductor/lowering.py]: Adopted the tighter stride[-2]-only structure you proposed (converting just the one relevant stride, guarded by len(stride) >= 2) — that's cleaner than converting the whole stride list, thanks.

I kept the symbolic-stride[-2] branch as a fall-back to check 1, not a raise. The reason is a recent PR in main added main's automatic-dynamic compilation: stride[-2] becomes a symbol (s50) for legitimate sliced copies whose offset still lands on a boundary and which the backend compiles correctly — e.g. a dynamic row narrow (stride=[s0, 1]) or the permute copy (stride=[64, 192, 1], offset 192, 192 % 192 == 0). Rejecting all symbolic strides regresses those.

**Comment by ani300** [comment]
@thoangtrvn will this be updated or closed?

**Comment by thoangtrvn** [comment]
We will come back to this as needed.
